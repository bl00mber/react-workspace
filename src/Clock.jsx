import React from 'react';

class Clock extends React.Component {
  constructor(props) {
    super(props);
    this.state = { date: new Date() };
  }

  componentDidMount() {
    // setting up the timer when component is rendered to DOM for the first time
    this.timerID = setInterval(
      () => this.tick(),
      1000
    );
  }

  componentWillUnmount() {
    // clearing the timer when component is removed from the DOM
    clearInterval(this.timerID);
  }

  tick() {
    this.setState({
      date: new Date()
    });
  }

  render() {
    return (
      <div>
        <h3>TimerComponent</h3>
        <p>It is {this.state.date.toLocaleTimeString()}.</p>
      </div>
    );
  }
}

export default Clock;
