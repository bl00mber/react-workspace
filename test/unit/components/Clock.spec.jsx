import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Clock from '@/components/Clock';

// mocking global Date object
const DATE_TO_USE = new Date(2018, 4, 15, 18, 50, 37);
jest.spyOn(DATE_TO_USE, 'toLocaleTimeString').mockReturnValue('18:50:37');
const mockedDate = Date;
global.Date = jest.fn(() => DATE_TO_USE);
global.Date.UTC = mockedDate.UTC;
global.Date.parse = mockedDate.parse;
global.Date.now = mockedDate.now;

jest.useFakeTimers();

describe('Clock', () => {
  test('should match enzyme shallow snapshot', () => {
    const wrapper = shallow(<Clock />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should call setInterval when component mounts', () => {
    const wrapper = shallow(<Clock />);
    const instance = wrapper.instance();
    instance.tick = jest.fn();
    setInterval.mockReset();
    setInterval.mockReturnValue('fnSetInterval');
    instance.componentDidMount();
    expect(instance.timerID).toBe('fnSetInterval');
    expect(setInterval).toHaveBeenCalled();
    expect(instance.tick).not.toHaveBeenCalled();
    jest.runOnlyPendingTimers();
    expect(instance.tick).toHaveBeenCalledWith();
  });

  test('should call clearInterval when component unmounts', () => {
    const wrapper = shallow(<Clock />);
    const instance = wrapper.instance();
    instance.tick = jest.fn();
    clearInterval.mockReset();
    instance.componentWillUnmount();
    expect(clearInterval).toHaveBeenCalledWith(instance.timerID);
  });

  test('tick should call setState with the date', () => {
    const wrapper = shallow(<Clock />);
    const instance = wrapper.instance();
    instance.setState = jest.fn();
    instance.tick();
    expect(instance.setState).toHaveBeenCalledWith({ date: DATE_TO_USE });
  });
});
