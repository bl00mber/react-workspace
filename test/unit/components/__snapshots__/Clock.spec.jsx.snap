// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock should match enzyme shallow snapshot 1`] = `
<div>
  <h3>
    TimerComponent
  </h3>
  <p>
    It is 
    18:50:37
    .
  </p>
</div>
`;
